File: msa/board-service/src/main/java/com/example/spring/boardservice/service/BoardService.java
package com.example.spring.boardservice.service;

import com.example.spring.boardservice.client.AuthClient;
import com.example.spring.boardservice.domain.entity.Board;
import com.example.spring.boardservice.domain.repository.BoardRepository;
import com.example.spring.boardservice.dto.BoardListItemResponseDto;
import com.example.spring.boardservice.dto.BoardSearchRequestDto;
import com.example.spring.boardservice.dto.UserNameResponseDto;
import com.example.spring.boardservice.exception.BoardNotFoundException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

import java.time.LocalDateTime;
import java.util.List;

@Slf4j
@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class BoardService {

    private final BoardRepository boardRepository;
    private final AuthClient authClient;
    private final FileService fileService;

    // repository는 userId까지만 채워서 돌려준다.
    // 페이지에 등장한 userId를 "모아서 한 번" auth에 요청(벌크)
    // 받은 DTO목록에서 이름을 찾아 채워 완성한다.
    public Page<BoardListItemResponseDto> searchBoards(BoardSearchRequestDto dto, Pageable pageable) {

        // searchBoards 게시글들 가져오기
        Page<BoardListItemResponseDto> page = boardRepository.searchBoards(dto, pageable);

        // boardRepository에서 가져온 ID추려서 auth-service로 요청해서 userName들 받아오기
        List<UserNameResponseDto> userNameResponseDtos = fetchNames(
                page.getContent().stream().map(BoardListItemResponseDto::getUserId).distinct().toList()
        );

        return page.map( item -> new BoardListItemResponseDto(
                item.getId(),
                item.getTitle(),
                item.getUserId(),
                userNameOf(userNameResponseDtos, item.getUserId()),
                item.getCommentCount(),
                item.getCreated()
        ));
    }

    public Board getBoardWithComments(Long boardId) {
        return boardRepository.findWithComments(boardId)
                .orElseThrow(
                        () -> new BoardNotFoundException("게시글을 찾을 수 없습니다. Id = " + boardId)
                );
    }

    // auth가 죽어도 게시판 조회 자체는 살아야 하므로(부분 실패 허용)
    // 실패 시 빈 목록을 돌려 이름 없이 응답한다. -> 장애 전파를 끊는다.
    private List<UserNameResponseDto> fetchNames(List<String> userIds) {

        if ( userIds == null || userIds.isEmpty() ) {
            return List.of();
        }

        try {
            return authClient.getUserNames(userIds);
        } catch (Exception e) {
            log.warn("[작성자 이름 조회 실패] auth-service 호출 불가 — userId로 대체 표시. {}", e.getMessage());
            return List.of();
        }

    }

    // DTO 목록에서 해당 userId의 이름을 찾는다. 없으면 null
    private String userNameOf(List<UserNameResponseDto> userNames, String userId) {
        return userNames.stream()
                .filter( userName -> userName.getUserId().equals(userId) )
                .map(UserNameResponseDto::getUserName)
                .findFirst()
                .orElse(null);
    }

    @Transactional
    public void saveBoard(String userId, String title, String content, MultipartFile file) {

        String filePath = fileService.storeFile(file);

        boardRepository.save(
                Board.builder()
                        .userId(userId)
                        .title(title)
                        .content(content)
                        .filePath(filePath)
                        .created(LocalDateTime.now())
                        .build()
        );

    }
}

File: kotlin/kotlin-part-4/src/main/kotlin/com/example/spring/kotlinpart4/service/BoardService.kt
package com.example.spring.kotlinpart4.service

import com.example.spring.kotlinpart4.domain.entity.Board
import com.example.spring.kotlinpart4.domain.repository.BoardRepository
import com.example.spring.kotlinpart4.dto.BoardCreateRequest
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class BoardService (private val repository: BoardRepository) {

    // 게시글 목록 및 검색 내역 조회
    fun getBoards(page: Int, size: Int, keyword: String?): Page<Board> {

        val pageable: Pageable = PageRequest.of(page - 1, size, Sort.by("id").descending())

        return if ( keyword.isNullOrBlank() ) {
            repository.findAll(pageable)
        } else {
            repository.findByTitleContaining(keyword, pageable)
        }
    }

    // 게시글 작성
    @Transactional
    fun createBoard(request: BoardCreateRequest) {
        repository.save( request.toEntity() )
    }

    // 게시글 상세 조회
    fun getBoard(id: Long): Board? = repository.findByIdOrNull(id)

}